### PR TITLE
Mihai week4 comment sentiment analysis

### PR DESCRIPTION
**Pull request by PredaMihaiDragos**
This pull requests uses the Google Sentiment Analysis API to give the comments a color from red to yellow to green, depending of how positive the comment is.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -142,4 +147,26 @@ public void doDelete(HttpServletRequest request, HttpServletResponse response) t
...
+      // Init doc element containing text
+      Document doc =
+        Document.newBuilder().setContent(text).setType(Document.Type.PLAIN_TEXT).build();
+
+      // Init the languageService
+      LanguageServiceClient languageService = LanguageServiceClient.create();
+    
+      // Analyze the doc's sentiment using the languageService and get a score
+      Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
+      float score = sentiment.getScore();
```
Comment (PredaMihaiDragos) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:192]: I am currently getting the sentiment score as a float (this is how the API returns it), saving it in datastore, and retrieving it as a Double (couldn’t find datastore support for floats). Is there a better approach for this? 
Also refer to lines 87 and 118

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 2]
```diff
@@ -19,31 +19,42 @@
...
+    this.sentimentScore = sentimentScore;
   }
 
-  public Comment(long id, String message, String addedBy) {
-    this(id, message, addedBy, new Date());
+  public Comment(long id, String message, String addedBy, Double sentimentScore) {
+    this(id, message, addedBy, new Date(), sentimentScore);
   }
 
+  // Getters
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: This comment is redundant.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (martemyev) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:172]: What if you cast it to Double here and make everything deal with the score as a double? There is no loss of precision when converting float to double.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -142,4 +147,26 @@ public void doDelete(HttpServletRequest request, HttpServletResponse response) t
...
+
+      // Close the languageService because we no longer need it
+      languageService.close();
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:195]: Do you need to close it explicitly? What happens with the service when the variable goes out of scope?

**Review by martemyev — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (saiavin) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:192]: Also, I think score must be recalculated every time the page is loaded? 
This is so, if the API changes its behavior it is reflected here as well.

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 2 above)
Comment (PredaMihaiDragos) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Thanks, deleted

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (PredaMihaiDragos) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:195]: It closes automatically when it goes out of scope, so I don't need to close it. But AFAIK it's the standard for good practice.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (PredaMihaiDragos) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:192]: > What if you cast it to Double here and make everything deal with the score as a double? There is no loss of precision when converting float to double.

Thanks for suggestion! I wanted to do that before, but I found out there is no straightforward way to convert from a float to a Double (you have to go through a string). Popular ways of converting are [here](http://programmingjungle.blogspot.com/2013/03/float-to-double-conversion-in-java.html#:~:text=There%20are%20a%20three%20popular,(new%20Float(fvar). ) FloatingDecimal class is no longer supported (can't import it) and the others look a little ugly to me, do you think I should do it like this? In my opinion letting the datastore class cast it looks cleaner

Comment (PredaMihaiDragos) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:192]: > Also, I think score must be recalculated every time the page is loaded?
> This is so, if the API changes its behavior it is reflected here as well.

That's a good point, thank you! I think it makes more sense to do that, but the [walkthrough](https://github.com/PredaMihaiDragos/Google-STEP/blob/master/walkthroughs/week-4-libraries/sentiment-analysis/sentiment-analysis-walkthrough.md) says to calculate it when the user enters it and to store it in the database and I just followed it... Do you think I should change the logic?

Comment (martemyev) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:192]: > > What if you cast it to Double here and make everything deal with the score as a double? There is no loss of precision when converting float to double.
> 
> Thanks for suggestion! I wanted to do that before, but I found out there is no straightforward way to convert from a float to a Double (you have to go through a string). Popular ways of converting are [here](http://programmingjungle.blogspot.com/2013/03/float-to-double-conversion-in-java.html#:~:text=There%20are%20a%20three%20popular,(new%20Float(fvar).) FloatingDecimal class is no longer supported (can't import it) and the others look a little ugly to me, do you think I should do it like this? In my opinion letting the datastore class cast it looks cleaner

I think it is fine the way it is implemented right now.

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -254,6 +261,68 @@ function deleteComment(commentId) {
...
+  return {
+    'r': interpolateValue(colorStart.r, colorStop.r, percentage),
+    'g': interpolateValue(colorStart.g, colorStop.g, percentage),
+    'b': interpolateValue(colorStart.b, colorStop.b, percentage),
+  };
+}
+
+/**
+ * Function that returns an interpolated value between start and stop based on percentage
+ * percentage is between [0, 1]
```
Comment (martemyev) [inline portfolio/src/main/webapp/script.js]: Nit: percentage is a misnomer for something in the [0,1] range. Maybe fraction or factor would be better.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (martemyev) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:195]: Well, let's agree to disagree :) To me, if the context gets closed (resource gets released) once a variable holding it goes out of scope, it is good practice to let the language do its thing. If you'd like to be explicit, that's okay, but the less code there is to read the better (to me).

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (PredaMihaiDragos) [reply portfolio/src/main/webapp/script.js]: Right, thank you! I changed it to fraction

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 5]
```diff
@@ -19,31 +19,40 @@
 /** Class containing a comment. */
 public final class Comment {
 
+  // Comment variables declaration
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: This is redundant.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 6]
```diff
@@ -19,31 +19,40 @@
...
   private final long id;
   private final String message;
   private final String addedBy;
   private final Date addedDate;
+  private final Double sentimentScore;
 
-  public Comment(long id, String message, String addedBy, Date addedDate) {
+  // Constructors
+
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: This is redundant.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 7]
```diff
@@ -19,31 +19,40 @@
...
     this.message = message;
     this.addedBy = addedBy;
     this.addedDate = addedDate;
+    this.sentimentScore = sentimentScore;
   }
 
-  public Comment(long id, String message, String addedBy) {
-    this(id, message, addedBy, new Date());
+  public Comment(long id, String message, String addedBy, Double sentimentScore) {
+    this(id, message, addedBy, new Date(), sentimentScore);
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: Is this constructor used at all?

**Review by martemyev — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 5 above)
Comment (PredaMihaiDragos) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Deleted

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 6 above)
Comment (PredaMihaiDragos) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Deleted

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 7 above)
Comment (PredaMihaiDragos) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Ah, I forgot to delete it, right! Thanks

**Review by martemyev — APPROVED** [review]

**Review by saiavin — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (saiavin) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:192]: > Do you think I should change the logic?

I think it's fine this way, if the walkthrough suggests that. But it's something you can keep in mind for future while storing results of API calls.

**Pull request merged**